### DISCUSSION TITLE
How to working with other existing database and tables

### DISCUSSION THREAD

**Discussion by Loclhh** [oqtane/oqtane.framework#1836] [Q&A] [answered]
I have a situation that I hope someone can guide me how to solve it. I have a central SQL server database that my company has been working with since many years and it has many tables. Is there any way that I can build an oqtane external module to work with a table on the existed database and **still remains** all oqtane capilities?

I think that by modifying the connection string that external module's dbcontext used when I create the external module to point to existed database then the module can work with a table on the database but I found no way to do that.

If anyone has experience with similar situations pls give me some advices or guides.

Thank in advance and nice day to all!

**Comment by hishamco** [comment, marked as answer]
You can create a `DbContext` that works with your exist database, or you could build a service layer and access what you need through APIs

**Reply by sbwalker** [reply]
@hishamco is correct with the 2 options he suggested above. If you want to use the DbContext option you would want to change the declaration in your module project from:

```
public class MyContext : DBContextBase, IService, IMultiDatabase
```
to
```
public class MyContext : DbContext, IService
```

and use the OnConfiguring method to wire up your existing database (most likely using DbContextOptionsBuilder and a named connection string in appsettings.json referenced via IConfiguration )

You would not want to inherit from DBContextBase or reference IMultiDatabase as these support the multi-tenant and multi-database aspects of the Oqtane framework itself. You should reference IService however as it provides the auto service registration capability during startup.

**Reply by Loclhh** [reply]
> @hishamco is correct with the 2 options he suggested above. If you want to use the DbContext option you would want to change the declaration in your module project from:
> 
> ```
> public class MyContext : DBContextBase, IService, IMultiDatabase
> ```
> 
> to
> 
> ```
> public class MyContext : DbContext, IService
> ```
> 
> and use the OnConfiguring method to wire up your existing database (most likely using DbContextOptionsBuilder and a named connection string in appsettings.json referenced via IConfiguration )
> 
> You would not want to inherit from DBContextBase or reference IMultiDatabase as these support the multi-tenant and multi-database aspects of the Oqtane framework itself. You should reference IService however as it provides the auto service registration capability during startup.


Base on Shaun Walker's instructions, I successes to build my module that can work with a table on existed database. I drop some detail notes here for other people can use it as need:

1. Change the context class: 
My code here: 
    //public class DigiBankContext : DBContextBase, IService, IMultiDatabase (this is the origin generated code line)
    public class DigiBankContext : DbContext, IService //change to this line
    {
        private readonly string _connectionString;
        private readonly IDatabase _database;
        private readonly IConfigurationRoot _config;
        public virtual DbSet<Models.DigiBank> DigiBank { get; set; }
       
        //Rem this default generated code
        //public DigiBankContext(ITenantManager tenantManager, IHttpContextAccessor accessor) : base(tenantManager, accessor)
        //{
        //    // ContextBase handles multi-tenant database connections
        //}

        public DigiBankContext(IConfigurationRoot config)
        {
            _config = config;
            //var connectionString = NormalizeConnectionString(_config.GetConnectionString(SettingKeys.ConnectionStringKey));  //defautl connection string
            var connectionString = NormalizeConnectionString(_config.GetConnectionString("MyExternalDatabaseConnectionString")); //chane to get my external database connection string
            var databaseType = _config.GetSection(SettingKeys.DatabaseSection)[SettingKeys.DatabaseTypeKey];

            IDatabase database = null;
            if (!string.IsNullOrEmpty(databaseType))
            {
                var type = Type.GetType(databaseType);
                database = Activator.CreateInstance(type) as IDatabase;
            }
            _connectionString = connectionString;
            _database = database;            
        }
        
        private string NormalizeConnectionString(string connectionString)
        {
            var dataDirectory = AppDomain.CurrentDomain.GetData("DataDirectory")?.ToString();
            connectionString = connectionString.Replace("|DataDirectory|", dataDirectory);
            return connectionString;
        }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
          => optionsBuilder.UseOqtaneDatabase(_database, _connectionString);
    }

2. Change Manager class:
    //public class DigiBankManager : MigratableModuleBase, IInstallable, IPortable (this is the origin generated code line)
    public class DigiBankManager : MigratableModuleBase, IPortable //change to this line
    {
        private IDigiBankRepository _DigiBankRepository;
        private readonly ITenantManager _tenantManager;
        private readonly IHttpContextAccessor _accessor;

        public DigiBankManager(IDigiBankRepository DigiBankRepository, ITenantManager tenantManager, IHttpContextAccessor accessor)
        {
            _DigiBankRepository = DigiBankRepository;
            _tenantManager = tenantManager;
            _accessor = accessor;
        }
        //Rem Install and Uninstall methods
        //public bool Install(Tenant tenant, string version)
        //{
        //    return Migrate(new DigiBankContext(_tenantManager, _accessor), tenant, MigrationType.Up);
        //}

        //public bool Uninstall(Tenant tenant)
        //{
        //    return Migrate(new DigiBankContext(_tenantManager, _accessor), tenant, MigrationType.Down);
        //}

        public string ExportModule(Module module)
        {
            string content = "";
            List<Models.DigiBank> DigiBanks = _DigiBankRepository.GetDigiBanks(module.ModuleId).ToList();
            if (DigiBanks != null)
            {
                content = JsonSerializer.Serialize(DigiBanks);
            }
            return content;
        }

        public void ImportModule(Module module, string content, string version)
        {
            List<Models.DigiBank> DigiBanks = null;
            if (!string.IsNullOrEmpty(content))
            {
                DigiBanks = JsonSerializer.Deserialize<List<Models.DigiBank>>(content);
            }
            if (DigiBanks != null)
            {
                foreach(var DigiBank in DigiBanks)
                {
                    _DigiBankRepository.AddDigiBank(new Models.DigiBank { ModuleId = module.ModuleId, CifBranch = DigiBank.CifBranch });
                }
            }
        }
    }

Now this module can work with the existed table in the external database.
I hope this will be helpful and save time for someone like me.

Cheer!

**Comment by lanthonyneville** [comment]
I can offer one other possible solution. However, this is not as technically "correct" as the one offered by @hishamco and @sbwalker .... I would suggest that you investigate that one first. My solution is MS SQL Server specific.

In this solution, you make use of a SQL Server feature called "synonyms". Creating a synonym creates a virtual object in your application database that is actually a reference to an object in a remote database. You can reference any type of object - tables, procedures, functions etc. After creating the synonym, you can work with it in exactly the same way that you would with an object that is physically in your application database. The remote database can be on another server, but performance will suffer if you run queries joining tables together.

To create a synonym, run: create synonym dbo.MySynonymName for [RemoteDbName].[dbo].[RemoteObjectName]

The advantage of this method is that all data access can be performed under one connection string. Also queries can join tables physically in the application DB with synonym'd tables with little performance penalty (in my experience, so long as the remote DB is in the same SQL Server).

**Comment by Loclhh** [comment]
Thank you very much  to  Hishamco, Sbwalker and Lanthonyneville. 
I'm very sorry that I couldn't reply your help soon because I got covid-19 and had to stay in bed more than 2 weeks. It was the terrible time to me that I feel very bad in my body despite the fact that I already vaccinated 2 injections before. 
I'm very appreciated with all information in your replies and I will build my module base on your instructions.
Thanks again and nice day to all!

**Reply by hishamco** [reply]
Sorry to hear that, hope to get better soon. Please let us know if you need any help
